### PR TITLE
Implement serde of Handle & GenericHandle

### PR DESCRIPTION
**Pull request by kabergstrom**
This PR implements the plumbing that enables the changes in my amethyst fork, please see this comparison and review: https://github.com/kabergstrom/amethyst/compare/e99835ac4c3aba2337f86c7f51407375f0ab50e6...new-asset-system

### CODE REVIEW

File: daemon/src/source_pair_import.rs [hunk 1]
```diff
@@ -182,74 +203,76 @@ impl<'a> SourcePairImport<'a> {
...
+                    ),
+                ));
+                let asset_data = &asset.asset_data;
+                let serialized_asset =
+                    SerializedAsset::create(asset_data.as_ref(), CompressionType::None, scratch_buf)?;
+                let build_pipeline = metadata
+                    .assets
+                    .iter()
+                    .find(|a| a.id == asset.id)
+                    .map(|m| m.build_pipeline)
```
Comment (Frizi) [inline daemon/src/source_pair_import.rs]: this looks like a place for `and_then`

File: importer/src/boxed_importer.rs [hunk 2]
```diff
@@ -145,3 +145,23 @@ pub fn get_source_importers(
...
+/// Use [inventory::submit!] to register an importer context to be entered for import operations.
+#[derive(Debug)]
+pub struct ImporterContextRegistration {
+    pub instantiator: fn() -> Box<dyn ImporterContext>,
+}
+inventory::collect!(ImporterContextRegistration);
+
+/// Get the registered importer contexts 
+pub fn get_importer_contexts(
+) -> impl Iterator<Item = Box<dyn ImporterContext + 'static>> {
```
Comment (Frizi) [inline importer/src/boxed_importer.rs]: Is it always `'static`? If so, then the trait should just have that as a bound.

File: loader/src/loader.rs [hunk 3]
```diff
@@ -4,7 +4,7 @@ use std::{error::Error, sync::Arc};
 
 /// Loading ID allocated by `atelier-assets` to track loading of a particular asset.
 #[derive(Copy, Clone, PartialEq, Eq, Debug, Hash)]
-pub struct LoadHandle(pub(crate) u64);
+pub struct LoadHandle(pub u64);
```
Comment (Frizi) [inline loader/src/loader.rs]: Why is this public? This allows fabricating fake load handles anywhere.

File: loader/src/rpc_loader.rs [hunk 4]
```diff
@@ -265,6 +265,18 @@ impl Loader for RpcLoader {
     }
 }
 
+impl LoaderInfoProvider for (&DHashMap<AssetUuid, LoadHandle>, &DHashMap<LoadHandle, AssetLoad>, &HandleAllocator) {
+    fn get_load(&self, id: AssetUuid) -> Option<LoadHandle> {
+        self.0.get(&id).map(|l| *l)
+    } 
+    fn get_asset_id(&self, load: &LoadHandle) -> Option<AssetUuid> {
```
Comment (Frizi) [inline loader/src/rpc_loader.rs]: reference to copy type `&LoadHandle`

File: loader/src/loader.rs [hunk 5]
```diff
@@ -203,3 +205,34 @@ pub trait Loader {
...
+    ///
+    /// * `id`: UUID of the asset.
+    fn get_load(&self, id: AssetUuid) -> Option<LoadHandle>;
+
+    /// Returns the AssetUUID for the given LoadHandle, if present.
+    ///
+    /// # Parameters
+    ///
+    /// * `load_handle`: ID allocated by `atelier-assets` to track loading of the asset.
+    fn get_asset_id(&self, load: &LoadHandle) -> Option<AssetUuid>;
```
Comment (Frizi) [inline loader/src/loader.rs]: reference to copy type `&LoadHandle`

File: loader/src/loader.rs [hunk 6]
```diff
@@ -68,13 +68,15 @@ pub trait AssetStorage {
...
     /// * `data`: The updated asset byte data.
     /// * `load_handle`: ID allocated by `atelier-assets` to track loading of a particular asset.
     /// * `load_op`: Allows the loading implementation to signal when loading is done / errors.
     /// * `version`: Runtime load version of this asset, increments each time the asset is updated.
     fn update_asset(
         &self,
+        loader_info: &dyn LoaderInfoProvider,
         asset_type_id: &AssetTypeId,
         data: &[u8],
         load_handle: &LoadHandle,
```
Comment (Frizi) [inline loader/src/loader.rs]: reference to copy type `&LoadHandle`

File: loader/src/rpc_loader.rs [hunk 7]
```diff
@@ -301,13 +313,30 @@ fn update_asset_metadata(
...
+            new_version: None,
+            asset_type: None,
+        }
+    }
+}
+
 fn load_data(
+    loader_info: &dyn LoaderInfoProvider,
     chan: &Arc<Sender<HandleOp>>,
     handle: &LoadHandle,
```
Comment (Frizi) [inline loader/src/rpc_loader.rs]: reference to copy type `&LoadHandle`

File: daemon/src/daemon.rs [hunk 8]
```diff
@@ -72,6 +75,20 @@ impl AssetDaemon {
...
+    pub fn with_importer_context(mut self, context: Box<dyn ImporterContext>) -> Self {
+        self.importer_contexts.push(context);
+        self
+    }
+
+    pub fn with_importer_contexts<I>(self, contexts: I) -> Self
+    where
+        I: IntoIterator<Item = Box<dyn ImporterContext>>,
+    {
+        contexts.into_iter().fold(self, |this, context| {
```
Comment (Frizi) [inline daemon/src/daemon.rs]: that seems to be cleaner with
```rust
self.importer_contexts.extend(contexts);
self
```

File: daemon/src/source_pair_import.rs [hunk 9]
```diff
@@ -174,6 +178,23 @@ impl<'a> SourcePairImport<'a> {
         Ok(())
     }
 
+    fn enter_import_contexts<F, R>(import_contexts: Option<&Vec<Box<dyn ImporterContext>>>, f: F) -> R
```
Comment (Frizi) [inline daemon/src/source_pair_import.rs]: Inconsistent naming. Is it `import_context` or `importer_context`?

File: daemon/src/source_pair_import.rs [hunk 10]
```diff
@@ -95,6 +96,9 @@ impl<'a> SourcePairImport<'a> {
         self.importer = importers.get_by_path(&self.source);
         Ok(self.importer.is_some())
     }
+    pub fn with_importer_contexts(&mut self, importer_contexts: &'a Vec<Box<dyn ImporterContext>>) {
```
Comment (Frizi) [inline daemon/src/source_pair_import.rs]: Those are not builder-style `with` methods, as those don't return `Self` so don't chain. I would name that `set_importer_contexts`. Same goes to `with_source_hash` and `with_meta_hash`.

File: loader/src/loader.rs [hunk 11]
```diff
@@ -203,3 +205,34 @@ pub trait Loader {
...
+pub trait LoaderInfoProvider {
+
+    /// Returns the load handle for the asset with the given UUID, if present.
+    ///
+    /// This will only return `Some(..)` if there has been a previous call to [`Loader::add_ref`].
+    ///
+    /// # Parameters
+    ///
+    /// * `id`: UUID of the asset.
+    fn get_load(&self, id: AssetUuid) -> Option<LoadHandle>;
```
Comment (Frizi) [inline loader/src/loader.rs]: That name reads awkwardly to me. How about `get_load_handle`?

File: loader/src/loader.rs (hunk 3 above)
Comment (kabergstrom) [reply loader/src/loader.rs]: It's for the serde "dummy" implementation, check the amethyst diff linked.

**Comment by kabergstrom** [comment]
Fixed, PTAL @azriel91 @Frizi

**Comment by kabergstrom** [comment]
And PTAL at https://github.com/kabergstrom/amethyst/compare/e99835ac4c3aba2337f86c7f51407375f0ab50e6...new-asset-system

File: loader/src/loader.rs (hunk 3 above)
Comment (Frizi) [reply loader/src/loader.rs]: I'd give it a pass, but it still smells bad :P Ideally the "mock" tools would simply live in this crate as well.

**Comment by azriel91** [comment]
Looks 99% good on that side, the [1%](https://github.com/kabergstrom/amethyst/compare/e99835ac4c3aba2337f86c7f51407375f0ab50e6...new-asset-system#diff-44f99085c296a2466c4903c3368214b2R203), thinking this should be an `.expect("")`, because the load handle should map to an `AssetUuid`. `Default::default()` would just be `[0u8; 16]` which makes us store incorrect information.

That's unrelated to here, so this side should happily merge.

**Pull request merged**